### PR TITLE
feature/1028_multitenancy_cartodbsink

### PR DESCRIPTION
**Pull request by frbattid**
* Implements issue #1028 
* 100% unit tests passed:
```
Tests run: 43, Failures: 0, Errors: 0, Skipped: 0 --> cygnus-common
Tests run: 91, Failures: 0, Errors: 0, Skipped: 0 --> cygnus-ngsi
```

Relevant tests for this PR:
```
[NGSICartoDBSink.start] ------------------------- Username field must appear within an entry in the keys file
[NGSICartoDBSink.start] ------------------  OK  - Null username has been detected
[NGSICartoDBSink.start] ------------------------- Username within an entry in the keys file cannot be empty
[NGSICartoDBSink.start] ------------------  OK  - Empty username has been detected
[NGSICartoDBSink.start] ------------------------- Endpoint within an entry in the keys file cannot be empty
[NGSICartoDBSink.start] ------------------  OK  - Empty endpoint has been detected
[NGSICartoDBSink.start] ------------------------- When started, a CartoDB backend is created from a valid keys file
[NGSICartoDBSink.start] ------------------  OK  - A CartoDB backend has been created
[NGSICartoDBSink.start] ------------------------- Key within an entry in the keys file cannot be empty
[NGSICartoDBSink.start] ------------------  OK  - Empty key has been detected
[NGSICartoDBSink.start] ------------------------- Key field must appear within an entry in the keys file
[NGSICartoDBSink.start] ------------------  OK  - Null key has been detected
[NGSICartoDBSink.start] ------------------------- Endpoint field must appear within an entry in the keys file
[NGSICartoDBSink.start] ------------------  OK  - Null endpoint has been detected
[NGSICartoDBSink.start] ------------------------- Endpoint within an entry in the keys file must be a URI using the http or https schema
[NGSICartoDBSink.start] ------------------  OK  - Invalid 'htt' schema in the endpoint has been detected
[NGSICartoDBSink.configure] --------------------- Configured `keys_conf_file` cannot be empty
[NGSICartoDBSink.configure] --------------  OK  - Empty 'keys_conf_file' was detected
```
* (unofficial) e2e tests passed:

Interesting logs at inititialization:
```
time=2016-05-20T12:00:35.273CEST | lvl=INFO | corr= | trans= | svc= | subsvc= | function=initializeBackend | comp=cygnus-ngsi | msg=com.telefonica.iot.cygnus.sinks.NGSICartoDBSink[151] : [distance-sink] Json containing CartoDB API keys has been read
time=2016-05-20T12:00:35.283CEST | lvl=INFO | corr= | trans= | svc= | subsvc= | function=initializeBackend | comp=cygnus-ngsi | msg=com.telefonica.iot.cygnus.sinks.NGSICartoDBSink[155] : [distance-sink] Json containing CartoDB API keys is syntactically OK
```

Received notitification:
```
time=2016-05-20T12:00:53.995CEST | lvl=INFO | corr=3fc4c457-ea8b-49a8-b010-12d79aad7a8e | trans=3fc4c457-ea8b-49a8-b010-12d79aad7a8e | svc=myuser | subsvc=/ | function=getEvents | comp=cygnus-ngsi | msg=com.telefonica.iot.cygnus.handlers.NGSIRestHandler[264] : Received data ({  "subscriptionId" : "51c0ac9ed714fb3b37d7d5a8",  "originator" : "localhost",  "contextResponses" : [    {      "contextElement" : {        "attributes" : [          {            "name" : "temperature",            "type" : "centigrade",            "value" : "26.5"          },          {            "name" : "the_geom",            "type" : "geometry",            "value" : "40.8104, -3.460002",            "metadatas": [              {                "name": "location",                "type": "string",                "value": "WGS84"              }            ]          }        ],        "type" : "Car",        "isPattern" : "false",        "id" : "Car1"      },      "statusCode" : {        "code" : "200",        "reasonPhrase" : "OK"      }    }  ]})
```

Persistence logs:
```
time=2016-05-20T12:01:07.416CEST | lvl=INFO | corr=3fc4c457-ea8b-49a8-b010-12d79aad7a8e | trans=3fc4c457-ea8b-49a8-b010-12d79aad7a8e | svc=myuser | subsvc=/ | function=persistDistanceEvent | comp=cygnus-ngsi | msg=com.telefonica.iot.cygnus.sinks.NGSICartoDBSink[568] : [distance-sink] Persisting data at NGSICartoDBSink. Schema (myuser), Table (car1_car_distance), Data ((1463738454013,'/','Car1','Car',(SELECT point FROM geom),(SELECT stage_distance FROM calcs),(SELECT stage_time FROM calcs),(SELECT stage_speed FROM speed),(SELECT sum_dist FROM inserts),(SELECT sum_time FROM inserts),(SELECT sum_speed FROM inserts),(SELECT sum2_dist FROM inserts),(SELECT sum2_time FROM inserts),(SELECT sum2_speed FROM inserts),(SELECT max_distance FROM inserts),(SELECT min_distance FROM inserts),(SELECT max_time FROM inserts),(SELECT min_time FROM inserts),(SELECT max_speed FROM inserts),(SELECT min_speed FROM inserts),(SELECT num_samples FROM inserts)))
time=2016-05-20T12:01:07.871CEST | lvl=INFO | corr=3fc4c457-ea8b-49a8-b010-12d79aad7a8e | trans=3fc4c457-ea8b-49a8-b010-12d79aad7a8e | svc=myuser | subsvc=/ | function=processNewBatches | comp=cygnus-ngsi | msg=com.telefonica.iot.cygnus.sinks.NGSISink[400] : Finishing internal transaction (3fc4c457-ea8b-49a8-b010-12d79aad7a8e)
```

* Assignee @pcoello25 but LGTM from @gtorodelvalle is also required

### CODE REVIEW

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java [hunk 1]
```diff
@@ -48,208 +56,295 @@ public NGSICartoDBSinkTest() {
...
         sink.configure(createContext(endpoint, apiKey, dataModel, enableLowercase, flipCoordinates, enableRaw,
-                enableDistance));
+                enableDistance, keysConfFile));
         
         try {
-            assertTrue(sink.invalidConfiguration);
+            assertTrue(sink.enableLowercase);
             System.out.println(getTestTraceHead("[NGSICartoDBSink.configure]")
-                    + "-  OK  - null value detected for 'endpoint'");
+                    + "-  OK  - 'enable_lowercase=false' was confiured, nevertheless it is always true by default");
```
Comment (pcoello25) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: confiured -> configured (Same in catch section)

**Comment by pcoello25** [comment]
LGTM, passing the ball to @gtorodelvalle 

@frbattid 1 - Tests 0 :laughing:

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java [hunk 2]
```diff
@@ -48,208 +56,295 @@ public NGSICartoDBSinkTest() {
...
         System.out.println(getTestTraceHead("[NGSICartoDBSink.configure]")
-                + "-------- Configured 'endpoint' must be a URI using the 'http' or 'https' schema");
-        String endpoint = "localhost:443";
+                + "-------- Configured 'flip_coordinates' cannot be different than 'true' or 'false'");
+        String endpoint = "https://localhost";
         String apiKey = "1234567890abcdef";
         String dataModel = null; // default one
         String enableLowercase = null; // default one
-        String flipCoordinates = null; // default one
+        String flipCoordinates = "falso";
```
Comment (gtorodelvalle) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Cervantes would be proud of us... :)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java [hunk 3]
```diff
@@ -48,208 +56,295 @@ public NGSICartoDBSinkTest() {
...
+                + "-------- Configured 'enable_raw' cannot be different than 'true' or 'false'");
         String endpoint = "https://localhost";
-        String apiKey = null;
+        String apiKey = "1234567890abcdef";
         String dataModel = null; // default one
         String enableLowercase = null; // default one
-        String flipCoordinates = null; // default one
-        String enableRaw = null; // default one
+        String flipCoordinates = null; // default value
+        String enableRaw = "falso";
```
Comment (gtorodelvalle) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Cervantes would be really proud of us... :)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java [hunk 4]
```diff
@@ -48,208 +56,295 @@ public NGSICartoDBSinkTest() {
...
+    public void testConfigureKeysConfFileOK() {
         System.out.println(getTestTraceHead("[NGSICartoDBSink.configure]")
-                + "-------- Configured `flip_coordinates` cannot be different than `true` or `false`");
+                + "-------- Configured 'keys_conf_file' cannot be empty");
         String endpoint = "https://localhost";
         String apiKey = "1234567890abcdef";
         String dataModel = null; // default one
         String enableLowercase = null; // default one
-        String flipCoordinates = "falso";
+        String flipCoordinates = null; // default one
```
Comment (gtorodelvalle) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: `default false` instead of `default one`?

Comment (gtorodelvalle) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Oh, I guess `null` is the default value, right?

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java (hunk 2 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: :) I'm really testing the sink configurator detects this `falso` value.

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java (hunk 3 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: :) I'm really testing the sink configurator detects this `falso` value.

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java (hunk 4 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Configuring something with `null` is the same than saying this parameter does not appear in the configuration. In this particular case, configuring `flip_coordinates` to null is the same than saying the configuration has no `flip_coordinates` parameter, thus, the `context.getString("flip_coordinates", "false")` gets the default value, which is `false`.

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java (hunk 1 above)
Comment (frbattid) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/1033/commits/948967e125859e6f768e8175076ab9ad1a815d9e

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java (hunk 2 above)
Comment (gtorodelvalle) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Ups! NTC :)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java (hunk 3 above)
Comment (gtorodelvalle) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: Ups! NTC :)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java (hunk 4 above)
Comment (gtorodelvalle) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSICartoDBSinkTest.java]: NTC :)

**Pull request merged**